src/persist-client/src/cli/inspect.rs

```diff
@@ -599,6 +543,247 @@ pub async fn unreferenced_blobs(
+        }
+    }
+
+    let mut current_parts = HashMap::new();
+    let mut current_rollups = HashSet::new();
+    state_iter.state().collections.trace.map_batches(|b| {
+        for part in b.parts.iter() {
+            current_parts.insert(
+                part.key.complete(&shard_id).to_string(),
+                part.encoded_size_bytes,
+            );
+        }
+    });
+    for rollup_key in state_iter.state().collections.rollups.values() {
+        current_rollups.insert(rollup_key.complete(&shard_id).to_string());
+    }
+
+    // There's a bit of a race condition between fetching s3 and state, but
+    // probably fine for most cases?
+    let (mut referenced_by_state_count, mut referenced_by_state_bytes) = (0, 0);
+    let (mut current_writer_pending_count, mut current_writer_pending_bytes) = (0, 0);
+    let (mut leaked_count, mut leaked_bytes) = (0, 0);
+    for (key, bytes) in s3_contents_before.iter() {
+        let referenced_by_state =
+            referenced_parts.contains_key(key) || referenced_rollups.contains(key);
+        if referenced_by_state {
+            referenced_by_state_count += 1;
+            referenced_by_state_bytes += bytes;
+        } else {
+            let current_writer_pending =
+                match BlobKey::parse_ids(key).expect("key should be in known format") {
+                    (_, PartialBlobKey::Batch(writer_id, _)) => state_iter
+                        .state()
+                        .collections
+                        .writers
+                        .contains_key(&writer_id),
+                    (_, PartialBlobKey::Rollup(_, _)) => false,
+                };
+            if current_writer_pending {
+                current_writer_pending_count += 1;
+                current_writer_pending_bytes += bytes;
+            } else {
+                leaked_count += 1;
+                leaked_bytes += bytes;
+            }
+        }
+    }
+    let (mut gc_able_parts_count, mut gc_able_parts_bytes) = (0, 0);
+    let (current_parts_count, current_parts_bytes) =
+        (current_parts.len(), current_parts.values().sum::<usize>());
+    for (key, bytes) in referenced_parts.iter() {
+        if !current_parts.contains_key(key) {
+            gc_able_parts_count += 1;
+            gc_able_parts_bytes += bytes;
+        }
+    }
+    let (mut gc_able_rollups_count, mut gc_able_rollups_bytes) = (0, 0);
+    let (mut current_rollups_count, mut current_rollups_bytes) = (0, 0);
+    for key in referenced_rollups.iter() {
+        let Some(bytes) = s3_contents_before.get(key) else {
+            println!("unknown size due to race condition: {}", key);
+            continue;
+        };
+        if current_rollups.contains(key) {
+            current_rollups_count += 1;
+            current_rollups_bytes += *bytes;
+        } else {
+            gc_able_rollups_count += 1;
+            gc_able_rollups_bytes += *bytes;
+        }
+    }
+
+    println!(
+        "total s3 contents: {} {} {}",
```

Comment:
unless you're pairwise reading the code, understanding the `{} {} {}` output is a little tricky at a glance. maybe add columns or write out the units?

Reply:
done! (we three hashed out the new format offline)